### PR TITLE
GUI evolution with event listeners

### PR DESCRIPTION
**Pull request by bottolo**
GUI for everyone!
- GUI has been added with a crosshair
- Event listeners are now implemented, and open up new possibilities in handling various things
- Added hints! For now, there's a hint which tells you to press the noclip button when in a `NoclipEnabler`, and one which tells you that you're not in a NoclipEnabler if you press such button
- (_Request_) Blue crosshair implemented

Known bugs:
- Noclip zone not found is shown when pressing P............................ in a `NoclipEnabler` :) but only when you're in _noclip mode_.

Thanks @spig95 for his help in the implementation. 
Also, the Parmigiana was good.

### CODE REVIEW

File: Assets/Code/Scripts/NoclipRealityManagement/NoclipManager.cs [hunk 1]
```diff
@@ -88,19 +89,27 @@ private IEnumerator DisableNoclip()
...
+            } 
+            else
+            {
+                EventManager.TriggerEvent("DisplayHint", "NOCLIP ZONE NOT FOUND"); 
+            }
+            
+            if (_playerCanDisableNoclip)
+            {
+                StartCoroutine(DisableNoclip());
+            }
```
Comment (spig95) [inline Assets/Code/Scripts/NoclipRealityManagement/NoclipManager.cs:112]:
```suggestion
            if (_playerCanDisableNoclip)
            {
                StartCoroutine(DisableNoclip());
            }
            else
            {
                EventManager.TriggerEvent("DisplayHint", "RETURN TO YOUR BODY TO EXIT NOCLIP");
            }
```

File: Assets/Code/Scripts/PlayerManagement/RealityPlayerCollisions.cs [hunk 2]
```diff
@@ -7,17 +8,22 @@ public class RealityPlayerCollisions : MonoBehaviour
...
+        _tutorialController = GetComponent<TutorialController>();
     }
 
     private void OnTriggerEnter(Collider other) {
         if (other.CompareTag("NoclipEnabler"))
         {
+            EventManager.TriggerEvent("PressP");
+            EventManager.TriggerEvent("DisplayHint", "PRESS P TO NOCLIP");
+            
             _noclipManager.SetPlayerCanEnableNoClip(true);
```
Comment (spig95) [inline Assets/Code/Scripts/PlayerManagement/RealityPlayerCollisions.cs]:
```suggestion
        if (other.CompareTag("NoclipEnabler"))
        {
            EventManager.TriggerEvent("DisplayHint", "PRESS P TO NOCLIP");
            _noclipManager.SetPlayerCanEnableNoClip(true);
```

File: Assets/Code/Scripts/TutorialManagement/TutorialController.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+    private void PressP()
+    {
+        EventManager.StopListening("PressP", PressP);
+        
+        Debug.Log("PressP displayed");
+        
+        EventManager.StartListening("PressP", PressP);
+    }
```
Comment (spig95) [inline Assets/Code/Scripts/TutorialManagement/TutorialController.cs]: At the end we did not use this function, feel free to keep it just to show an example of an event without inputs from the trigger. If you keep it, I would refactor it as follows
```suggestion
    private void Awake()
    {
        EventManager.StartListening ("PressP", PressP);
        EventManager.StartListening ("DisplayHint", DisplayHint);
    }


    private void PressP()
    {
        EventManager.StopListening("PressP", PressP);
        StartCoroutine(DisplayHintCoroutine("PRESS P"));
        EventManager.StartListening("PressP", PressP);
    }
```

File: Assets/Code/POLIMIgameCollective/EVENTM~1/EventManager.cs [hunk 4]
```diff
@@ -0,0 +1,217 @@
+﻿using UnityEngine;
```
Comment (spig95) [inline Assets/Code/POLIMIgameCollective/EVENTM~1/EventManager.cs]: I think that by mistake the folder has a weird name. Can you please change the "EVENTM-1" name? Thanks :)

**Review by spig95 — COMMENTED** [review]
I have left a couple of minor comments, but the event system works and it's really nice to have!

File: Assets/Code/Scripts/NoclipRealityManagement/NoclipManager.cs (hunk 1 above)
Comment (spig95) [inline Assets/Code/Scripts/NoclipRealityManagement/NoclipManager.cs]: Something like this should fix the bug and also give an additional and useful hint to the user. **Please try it, as I did not have time to test it.**
 
```suggestion
            if (_playerCanEnableNoclip)
                StartCoroutine(EnableNoclip());
            else if (_playerCanDisableNoclip)
                StartCoroutine(DisableNoclip());
            else if (!_noclipEnabled)
                EventManager.TriggerEvent("DisplayHint", "NOCLIP ZONE NOT FOUND. PRESSING P HAS NO EFFECT"); 
            else if (_noclipEnabled)
                EventManager.TriggerEvent("DisplayHint", "RETURN TO YOUR BODY TO DISABLE NOCLIP"); 
```

File: Assets/Code/POLIMIgameCollective/EVENTM~1/EventManager.cs (hunk 4 above)
Comment (bottolo) [reply Assets/Code/POLIMIgameCollective/EVENTM~1/EventManager.cs]: On it

**Comment by bottolo** [comment]
Now everything works as intended

**Review by TizioMaurizio — APPROVED** [review]
Hints are good but I would make press p to noclip disappear as soon as you are out of the platform, if easy to implement

**Pull request merged**